RUN-3286 App assets API does not work if not launched by manifest

wenjunche opened the pull request:
Test results:
[Windows 10](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/599f65a4d2a02971da3a6347
)
[Windows 7](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/599f6542d2a02971da3a6346)

Also, fixed code in authorization.js according to [https://github.com/openfin/openfin/wiki/RVM-Protocol]().  created RUN-3290 to track the issue.

New versions of .net and Java adapters are required to test app asset APIs from external adapters.

StevenEBarbaro commented:
👍

HarsimranSingh commented:
PR title format `RUN-#### <short description>` (no colon after the number)

whyn07m3 commented on src/browser/api_protocol/api_handlers/authorization.js:
```diff
@@ -99,7 +99,8 @@ function onRequestAuthorization(id, data) {
     const externalConnObj = Object.assign({}, data.payload, {
-        id
+        id,
+        configUrl: authObj.authReqPayload && authObj.authReqPayload.configUrl
```
This can maybe end up being `configUrl: null` or other falsy values. Will it brake somewhere down the line then? Maybe be should be this instead: 
```
const externalConnObj = Object.assign({}, data.payload, { id });

if (authObj.authReqPayload) {
   externalConnObj.configUrl = authObj.authReqPayload.configUrl;
}
```

wenjunche replied:
added check for authObj

whyn07m3 commented on src/browser/api_protocol/api_handlers/authorization.js:
```diff
@@ -123,14 +124,14 @@ function onRequestAuthorization(id, data) {
-                licenseKey: externalConnObj.licenseKey,
-                client: externalConnObj.client,
+                licenseKey: authObj.authReqPayload && authObj.authReqPayload.licenseKey,
+                client: authObj.authReqPayload && authObj.authReqPayload.client,
```
Same in these case as in the comment above...?

wenjunche replied:
Added check for authObj

whyn07m3 commented on src/browser/core_state.ts:
```diff
@@ -516,6 +521,15 @@ export function getAppAncestor(descendantAppUuid: string): Shapes.App {
     }
 }
 
+function getExternalAncestor(descendantAppUuid: string): any {
```
Can we specify shape here instead of `any`?

wenjunche replied:
We need to create shapes for external connection

whyn07m3 commented on src/browser/core_state.ts:
```diff
@@ -211,7 +211,12 @@ export function getUuidBySourceUrl(sourceUrl: string): string|boolean {
+        return app._configUrl;
+    } else {
+        const externalApp = getExternalAncestor(uuid);
+        return externalApp && externalApp.configUrl;
```
Allowed return values are `string|boolean`. This would have to be `return !!externalApp && externalApp.configUrl;`

nicholasdgoodman replied:
this function should not be returning `string|boolean` and falsey values should not be coerced to boolean. instead they should be coerced to `undefined` via:

```
// if getExternalAncestor is written correctly, this coercion isn't necessary.
const externalApp = getExternalAncestor(uuid) || {};

// returns a string value or undefined
return externalApp && externalApp.configUrl; 
```

whyn07m3 replied:
I like this better 👍

whyn07m3 replied:
... typescript reflects what we've had, and string|boolean is what we've had. Unless we start refactoring throughout, just safer to stick to what's there now?

whyn07m3 commented:
Also, we need to test this

wenjunche commented:
Added following check in authenticateUuid

if (!authObj) {
        cb(false, 'Invalid UUID: ' + authRequest.uuid);

So onRequestAuthorization returns an error if there is no pending connection for the UUID.

Modified Java adapter to test the condition (sending request-authorization without request-external-authorization)
